### ISSUE TITLE
Torch Elastic - How to make sure all nodes are in the same AZ?

### ISSUE THREAD

**Issue by thecooltechguy** [pytorch/elastic#140]
### Please note that this issue tracker is not a help form and this issue will be closed.

Before submitting, please ensure you have gone through our documentation. Here
are some links that may be helpful:

* [What is torchelastic?](../../README.md)
* [Quickstart on AWS](../../aws/README.md)
* [Usage](../../USAGE.md)
* [Examples](../../examples/README.md)
* API documentation
    * [Overview](../../USAGE.md)
    * [Rendezvous documentation](../../torchelastic/rendezvous/README.md)
    * [Checkpointing documentation](../../torchelastic/checkpoint/README.md)
* [Configuring](../../USAGE.md#configuring)

  
### Question

Hi, when using TorchElastic + AWS EKS, how can we ensure that multi-node training jobs have all of the nodes located in the same AZ? This is critical for multi-node training jobs, in terms of speed of data transfer and data transfer costs.

One naive way would be to just specify 1 subnet when creating the EKS cluster, but is there a way we can create an EKS cluster with multiple subnets, and when TorchElastic attempts to launch multiple nodes for a training job, it will try to launch them such that all of the nodes are located within 1 subnet/AZ (where that subnet would be one of the subnets that the EKS cluster has)? And is this possible to do with spot instances?

Thanks!

**Comment by kiukchung** [comment]
I believe this is possible in EKS by specifying multiple subnets on the VPC into which the EKS cluster launches. This doc may be helpful: https://docs.aws.amazon.com/eks/latest/userguide/eks-networking.html.

cc) @Jeffwan for question regarding pod packing.

**Comment by Jeffwan** [comment]
There're multiple ways to achieve this. 

1. AWS EKS just give you the control plane, you can add a node group within single AZ to the cluster. The workloads automatically fit into same AZ. 

2. In order to have high availability, some users still want to have multi-AZ data plane (single ASG). In this case, you need to leverage scheduling features. Either use nodeSelector or nodeAffinity to [assign pods to target nodes](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/) with same AZ label.


3. #2 will out of box working if you have enough nodes. If you use autoscaler and want to dynamically adding new spot nodes.  it won't work automatically. The major reason if you have multi-AZ in single AZ.  ASG will try to rebalance nodes, that means you won't have 1 node in AZ-1, 1 node in AZ-2, 5nodes in AZ-3 (where the job runs). In this case, you can either suspend AZ rebalancing or create 3 ASGs and make each use one AZ instead. 


If you still have questions on autoscaling, spot , check the best practice I wrote to get more ideas. https://aws.github.io/aws-eks-best-practices/cluster-autoscaling/cluster-autoscaling/
